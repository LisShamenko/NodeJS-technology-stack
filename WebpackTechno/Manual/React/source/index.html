<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <title>Плагин 'html-webpack-plugin'.</title>
</head>

<body>
    <div id="root"></div>
</body>

</html>