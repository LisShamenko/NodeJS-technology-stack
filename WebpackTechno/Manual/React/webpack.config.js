// --------------- 2. React.

// Для транспиляции кода с компонентами React требуется добавить предустановку 
//      babel для React.
//      npm i @babel/core babel-loader @babel/preset-env @babel/preset-react --save-dev

// Файл настройки babel для использования предустановки React.
//      {
//          "presets": [
//              "@babel/preset-env", 
//              "@babel/preset-react"
//          ]
//      }

// Webpack сам добавляет ссылку на файл bundle.js в index.html.

const path = require('path');
const HtmlWebpackPlugin = require("html-webpack-plugin");

module.exports = {
    mode: 'development',
    entry: {
        index: path.resolve(__dirname, "source", "index.js")
    },
    output: {
        path: path.resolve(__dirname, "build"),
        filename: 'bundle.js'
    },
    plugins: [
        new HtmlWebpackPlugin({
            template: path.resolve(__dirname, "source", "index.html")
        })
    ],
    module: {
        rules: [
            {
                test: /\.css$/,
                use: ["style-loader", "css-loader"]
            },
            {
                test: /\.scss$/,
                use: ["style-loader", "css-loader", "sass-loader"]
            },
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: {
                        presets: [
                            '@babel/preset-env',
                            "@babel/preset-react"
                        ],
                        plugins: [
                            "@babel/plugin-proposal-class-properties",
                            "@babel/plugin-transform-runtime"
                        ]
                    }
                }
            }
        ]
    },
};