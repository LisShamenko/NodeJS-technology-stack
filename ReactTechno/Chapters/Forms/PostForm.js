// --- 8. Работа с формами в React.

// Пользователи должны видеть предупреждение, когда они нарушают правила валидации 
//      вводимых данных, в этом случае валидация должна выполняться в браузере, 
//      чтобы обеспечить плавность UI и дополнительный уровень защиты. Аналогичная 
//      валидация должна выполняться и на сервере, так как сервер не может полагаться

// Для сетевых запросов используется библиотека 'isomorphic-fetch', которая 
//      использует API Fetch браузера, а также работает на сервере.

// импорт
import React from 'react';
import PropTypes from 'prop-types';

// пакет 'bad-words' позволяет фильтровать слова
import Filter from 'bad-words';

// компонент отвечает за создание постов
class PostForm extends React.Component {
    constructor(props) {
        super(props);

        // --- обработчики: привязка

        // обработчик для отправки формы
        this.handleSubmit = this.handleSubmit.bind(this);

        // обработчик для валидации формы
        this.handlePostChange = this.handlePostChange.bind(this);

        // установить начальное состояние
        this.state =
        {
            // изменение состояния компонента 
            content: '',

            // --- валидация данных

            // Разработка валидации данных форм должна включать следующие этапы:
            // 
            //      1. Следует определить требования к данным, что поможет определить  
            //          основные рекомендации по исправлению данных.
            //      2. Разработка компонентов с проверкой данных, которые позволят  
            //          пользователям предоставлять значимые данные.
            //      3. Определить способы устранения несоответствий в данных,  
            //          предоставляемых пользователями.

            // флаг валидации
            isValid: false,
            isControl: true,
        };

        // фильтр слов
        this.filter = new Filter();
    }

    // обработчик для валидации формы, обработчики событий получают событие 
    //      в качестве аргумента, событие компонента React является оберткой 
    //      для браузерного события
    handlePostChange(event) {

        console.log('--- handlePostChange --- start');

        // --- обработчики: event.target

        // event.target - это ссылка на DOM-элемент, который отправил событие, 
        //      аналогично обычному браузерному коду.
        console.log(`
            --- event = ${event} 
            --- event.target = ${event.target}
        `);

        // Работа с формами в React заключается в получении событий в обработчиках и 
        //      использование этих событий для обновления данных компонента. 

        // state и props - это два основных способа, с помощью которых React 
        //      позволяет работать с данными.

        // Одна из основных задач React заключается в синхронизации DOM браузера 
        //      с виртуальной DOM, которая собирается из компонентов. Таким образом
        //      при вводе в текстовое поле, символы не отображаются сразу, так как
        //      React должен сначала изменить виртуальную DOM. Вместо этого, 
        //      события ввода отправляются на обработку в React, где в обработчиках 
        //      принимается решение об изменении данных, меняется состояние компонента,
        //      после чего React синхронизирует виртуальную DOM с фактической. 
        //      
        //      ввод символа => генерация события => получение события        =>
        //      в textarea         onChange          в обработчике компонента
        //      -----------------------------------------------------------------
        //      изменение state => React меняет DOM браузера
        //      -----------------------------------------------------------------

        // значение value из DOM элемента textarea
        const content = event.target.value;

        // --- валидация данных: 3. устранение несогласованности данных

        // фильтрация
        let clear_content;
        try {
            clear_content = this.filter.clean(content);
        }
        catch (err) {
            console.log(`--- error of 'bad-words' --- error: ${err}`);
            clear_content = content;
        }

        // обновить this.state с учетом фильтра и с флагом валидации
        this.setState(() => {
            return {
                // изменение состояния компонента 
                content: clear_content,

                // --- валидация данных: 1. определить ограничения данных

                // правило валидации: контент не должен превышать заданной длины
                isValid: clear_content.length <= 280
            };
        });

        //
        console.log('--- handlePostChange --- end');
    }



    // обработчик для отправки формы
    handleSubmit(event) {

        console.log(`--- handleSubmit --- start`);

        // --- валидация данных: 2. выполнить валидацию данных

        // предотвратить событие 'submit' по умолчанию
        event.preventDefault();

        // проверка валидации
        if (!this.state.isValid) {
            return;
        }

        // --- создание сообщений

        // убедится в наличии функции onSubmit
        if (this.props.onSubmit) {

            // объект с данными нового поста
            const newPost = {
                content: this.state.content,
                userId: this.props.user.id,
            };

            // обратный вызов onSubmit:
            //      дочерний компонент несет ответственность за передачу
            //      данных в родительский компонент, чтобы тот мог передать 
            //      их на сервер

            // отправить данные нового поста в родительский компонент через
            //      функцию обратного вызова
            this.props.onSubmit(newPost);

            // сброс состояния компонента для создания следующего поста
            this.setState({
                content: '',
                isValid: null,
            });
        }

        //
        console.log(`--- handleSubmit --- end --- state = ${this.state}`);
    }



    // 
    render() {

        if (this.state.isControl) {

            // --- использование контролируемых компонентов

            // React позволяет отслеживать события как в обычном браузерном коде, независимо 
            //      от версий различных браузеров, то есть не нужно беспокоиться о различиях 
            //      между браузерами.

            // Отслеживаются два события:
            // - onChange, срабатывает при изменении ввода, новое значение элемента формы
            //      можно получить при помощи выражения 'event.target.value'
            // - onClick, срабатывает при щелчке мыши на элементе, что позволяет определить
            //      момент нажатия на кнопку и отправить сообщение на сервер

            return (
                // обработчики событий передаются в компоненты: button и textarea
                <div>
                    <button onClick={this.handleSubmit}>Post</button>
                    <textarea
                        // содержимое DOM элемента textarea определяется значением 
                        //      из this.state
                        value={this.state.content}
                        onChange={this.handlePostChange}
                    />
                </div>
            );
        }
        else {
            
            // --- использование неконтролируемых компонентов

            return (
                <div>
                    <button onClick={this.handleSubmit}>Post</button>
                    <textarea
                        // содержимое DOM элемента textarea не контролируется React
                        onChange={this.handlePostChange}
                    />
                </div>
            );
        }
    }
}

// валидация свойств
PostForm.propTypes = {

    // --- создание сообщений

    // определяется свойство, которому передается функция обратного вызова
    //      для передачи данных в родительский компонент
    onSubmit: PropTypes.func.isRequired,
    user: PropTypes.object,
}

// экспорт
export default PostForm;