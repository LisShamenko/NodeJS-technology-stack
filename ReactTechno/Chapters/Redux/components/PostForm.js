import React from 'react';
import PropTypes from 'prop-types';
import Filter from 'bad-words';

// 
class PostForm extends React.Component {

    // 
    constructor(props) {
        super(props);

        // 
        this.state =
        {
            content: '',
            isValid: false,
        };

        // 
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handlePostChange = this.handlePostChange.bind(this);
        this.filter = new Filter();
    }

    // 
    handlePostChange(event) {

        // 
        let clear_content;
        try {
            clear_content = this.filter.clean(event.target.value);
        }
        catch (err) {
            clear_content = event.target.value;
        }

        // 
        this.setState(() => {
            return {
                content: clear_content,
                isValid: clear_content.length <= 280
            };
        });
    }

    // 
    handleSubmit(event) {
        event.preventDefault();

        // 
        if (!this.state.isValid) {
            return;
        }

        // 
        if (this.props.onSubmit) {

            this.props.onSubmit({
                content: this.state.content,
                userId: this.props.user.id,
            });

            // 
            this.setState(() => {
                return {
                    content: '',
                    isValid: null,
                };
            });
        }
    }

    // 
    render() {
        return (
            <div>
                <button onClick={this.handleSubmit}>Post</button>
                <textarea
                    value={this.state.content}
                    onChange={this.handlePostChange}
                />
            </div>
        );
    }
}

// 
PostForm.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    user: PropTypes.object,
}

// 
export default PostForm;