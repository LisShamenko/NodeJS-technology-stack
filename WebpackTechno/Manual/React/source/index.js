import React, { useState, Fragment } from "react";
import { render } from "react-dom";



// --------------- 2. CSS в React.

import "./style.css";
import "./style.scss";

// Способы применения CSS в React:
//      - CSS-in-CSS (CSS, Sass, Модули CSS, Модули CSS и Sass)
//      - CSS-in-JS (Styled Components, Emotion)
//      - Utility-First-CSS (Tailwind CSS)

import Basket_base from "./components/basket_base";
import Basket_file_css from "./components/basket_file_css";
import Basket_file_scss from "./components/basket_file_scss";
import Basket_module_css from "./components/basket_module_css";
import Basket_styled_components from "./components/basket_styled_components";
import Basket_inline_css from "./components/basket_inline_css";

// 
function App() {

    // 
    const [state, setState] = useState("CLICK ME");

    // 
    const [fruits, setFruits] = React.useState([
        { id: '1', name: 'Apple', isFavorite: false },
        { id: '2', name: 'Peach', isFavorite: true },
        { id: '3', name: 'Strawberry', isFavorite: false },
    ]);

    //
    function handleClick(item) {
        const newFruits = fruits.map((fruit) => {
            if (fruit.id === item.id) {
                return {
                    id: fruit.id,
                    name: fruit.name,
                    isFavorite: !fruit.isFavorite,
                };
            }
            else {
                return fruit;
            }
        });
        setFruits(newFruits);
    }

    // 
    return (
        <Fragment>
            <div>
                <h1 className="html">Плагин HTML.</h1>
                <h1 className="sass">Загрузчик SASS.</h1>
                <button onClick={() => setState('click')} >
                    {state}
                </button>
                <hr />
            </div>
            <div>
                <h3>No style.</h3>
                <Basket_base items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Файл CSS.</h3>
                <Basket_file_css items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Файл SCSS.</h3>
                <Basket_file_scss items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Модульный CSS.</h3>
                <Basket_module_css items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Библиотека styled-components.</h3>
                <Basket_styled_components items={fruits} onClick={handleClick} />
                <hr />
            </div>
            <div>
                <h3>Встроенные стили React.</h3>
                <Basket_inline_css items={fruits} onClick={handleClick} />
                <hr />
            </div>
        </Fragment>
    );
}

// 
render(<App />, document.getElementById("root"));