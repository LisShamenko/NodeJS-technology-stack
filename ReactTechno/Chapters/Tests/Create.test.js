import React from 'react';

// выполняет рендеринг компонентов React для тестирования, 
//      поставляется вместе с Jest:
//      https://ru.reactjs.org/docs/test-renderer.html
import renderer from 'react-test-renderer';

// 
import PostForm from './../../components/Post/PostForm';



// --------------- 10. Тестирование.

// 
describe('component PostForm', () => {

    // 
    let user = {
        "id": "00000000-0000-0000-0000-000000000000",
        "name": "name",
        "image": "/static/images/users/0.jpg"
    };

    // 
    test('snapshot', () => {

        // jest.fn создает mock-функцию, которая позволяет изолировать код и делать
        //      утверждения о том, как компонент использовал mock-функцию, была ли она 
        //      вызвана и с какими аргументами
        const props = { onSubmit: jest.fn(), user: user };

        // чтобы протестировать конструктор следует использовать Sinon, в конструкторе 
        //      происходит инициализация состояния компонента, привязка методов и 
        //      другие настройки

        // создать обертку компонента
        const component = renderer.create(<PostForm {...props} />);

        // генерировать снимок компонента
        const tree = component.toJSON();

        // проверка снимка
        expect(tree).toMatchSnapshot();
    });

    // 
    test('handlePostChange', () => {

        // фиктивные объекты
        const props = { onSubmit: jest.fn(), user: user };
        const mockEvent = { target: { value: 'value' } };

        // имитация функции setState для обновления состояния компонента
        PostForm.prototype.setState = jest.fn(function (updater) {
            this.state = Object.assign(this.state, updater(this.state));
        });

        // создать компонент и вызвать тестируемый метод
        const component = new PostForm(props);
        component.handlePostChange(mockEvent);

        // проверки после выполнения тестируемого метода
        expect(component.setState).toHaveBeenCalled();
        expect(component.setState.mock.calls.length).toEqual(1);
        expect(component.state).toEqual({
            isValid: true,
            content: mockEvent.target.value,
        });
    });

    // 
    test('handleSubmit', () => {

        // фиктивные свойства
        const props = { onSubmit: jest.fn(), user: user };

        // фиктивный объект события
        const mockEvent = {
            target: { value: 'value' },
            preventDefault: jest.fn()
        };

        //
        PostForm.prototype.setState = jest.fn(function (updater) {
            this.state = Object.assign(this.state, updater(this.state));
        });

        //
        const component = new PostForm(props);

        // напрмую изменить состояние компонента и вызвать тестируемый метод
        component.state = {
            isValid: true,
            content: 'content'
        };
        component.handleSubmit(mockEvent);

        // проверка
        expect(component.setState).toHaveBeenCalled();
        expect(props.onSubmit).toHaveBeenCalledWith({
            content: 'content',
            userId: user.id
        });
    });
});