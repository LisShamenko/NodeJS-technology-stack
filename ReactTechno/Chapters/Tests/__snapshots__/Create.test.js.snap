// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component PostForm snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Create Post.
  </button>
  <textarea
    onChange={[Function]}
    value=""
  />
</div>
`;
